xarray/core/indexing.py

```diff
@@ -434,7 +492,84 @@ def __setitem__(self, key, value):
         self.array[key] = value
 
 
-def orthogonally_indexable(array):
+def unbroadcast_indexes(key, shape):
+    """
+    Convert broadcasted indexers to orthogonal indexers.
+    If there is no valid mapping, raises IndexError.
+
+    key is usually generated by Variable._broadcast_indexes.
+
+    key: tuple of np.ndarray, slice, integer
+    shape: shape of array
+    """
+    key = expanded_indexer(key, len(shape))
+
+    if all(isinstance(k, integer_types + (slice,)) for k in key):
+        # basic indexing
+        return key
+
+    # Here, we need to distinguish integer and 1-element array
```

Comment:
OK. Thanks for the comment.

> I wonder if instead of converting int/slice objects into xarray.Variable object for broadcasting, we could keep track of the original objects in some way.

It sounds cleaner to me.

I guess more backends would support the fancy indexing in the future?
If it is the major trend, I think it would be great to prepare logic to quickly adapt their future improvement.